sled-storage/src/resources.rs

```diff
@@ -203,4 +228,333 @@ impl StorageResources {
             .map(|p| p.join(BUNDLE_DIRECTORY).join(ZONE_BUNDLE_DIRECTORY))
             .collect()
     }
+
+    /// Returns an iterator over all managed disks.
+    pub fn iter_managed(&self) -> impl Iterator<Item = (&DiskIdentity, &Disk)> {
+        self.values.iter().filter_map(|(identity, disk)| match disk {
+            ManagedDisk::ExplicitlyManaged(disk) => Some((identity, disk)),
+            ManagedDisk::ImplicitlyManaged(disk) => Some((identity, disk)),
+            _ => None,
+        })
+    }
+
+    /// Returns an iterator over all disks, managed or not.
+    pub fn iter_all(
+        &self,
+    ) -> impl Iterator<Item = (&DiskIdentity, DiskVariant, i64)> {
+        self.values.iter().map(|(identity, disk)| match disk {
+            ManagedDisk::ExplicitlyManaged(disk) => {
+                (identity, disk.variant(), disk.slot())
+            }
+            ManagedDisk::ImplicitlyManaged(disk) => {
+                (identity, disk.variant(), disk.slot())
+            }
+            ManagedDisk::Unmanaged(raw) => {
+                (identity, raw.variant(), raw.slot())
+            }
+        })
+    }
+}
+
+/// The intersection of "physical disks noticed by hardware" and "physical
+/// disks requested by the control plane".
+#[derive(Debug)]
+pub struct StorageResources {
+    log: Logger,
+
+    key_requester: StorageKeyRequester,
+
+    // All disks, real and synthetic, that exist within this sled
+    disks: AllDisks,
+
+    // The last set of disks the control plane explicitly told us to manage.
+    //
+    // Only includes external storage (U.2s).
+    control_plane_disks: BTreeMap<DiskIdentity, OmicronPhysicalDiskConfig>,
```

Comment:
In the limit, I think yes, but I'm currently holding onto that information in `sled-storage/src/manager.rs` for now. It's weird to handle this case for "ledgerless" systems, because there is no known Generation value.